packages/dds/tree/src/feature-libraries/modular-schema/genericFieldKind.ts

```diff
@@ -106,57 +63,72 @@ export const genericChangeHandler: FieldChangeHandler<GenericChangeset> = {
 };
 
+function compose(
+	change1: GenericChangeset,
+	change2: GenericChangeset,
+	composeChildren: NodeChangeComposer,
+): GenericChangeset {
+	const composed = change1.clone();
+	for (const [index, id2] of change2.entries()) {
+		const id1 = composed.get(index);
+		const idComposed = id1 !== undefined ? composeChildren(id1, id2) : id2;
+		composed.set(index, idComposed);
+	}
+
+	return composed;
+}
+
 function getNestedChanges(change: GenericChangeset): [NodeId, number | undefined][] {
-	return change.map(({ index, nodeChange }) => [nodeChange, index]);
+	return change.toArray().map(([index, nodeChange]) => [nodeChange, index]);
 }
 
 function rebaseGenericChange(
 	change: GenericChangeset,
 	over: GenericChangeset,
 	rebaseChild: NodeChangeRebaser,
 ): GenericChangeset {
-	const rebased: GenericChangeset = [];
-	let iChange = 0;
-	let iOver = 0;
-	while (iChange < change.length || iOver < over.length) {
-		const a = change[iChange];
-		const b = over[iOver];
-		const aIndex = a?.index ?? Infinity;
-		const bIndex = b?.index ?? Infinity;
-		let nodeChangeA: NodeId | undefined;
-		let nodeChangeB: NodeId | undefined;
+	const rebased: GenericChangeset = change.clone();
+	let nextIndex = 0;
+
+	// eslint-disable-next-line no-constant-condition
+	while (true) {
+		const newEntry = change.getPairOrNextHigher(nextIndex);
+		const baseEntry = over.getPairOrNextHigher(nextIndex);
+
+		if (baseEntry === undefined && newEntry === undefined) {
+			break;
+		}
+
+		const newIndex = newEntry?.[0] ?? Infinity;
+		const baseIndex = baseEntry?.[0] ?? Infinity;
+		let newNodeChange: NodeId | undefined;
+		let baseNodeChange: NodeId | undefined;
 		let index: number;
-		if (aIndex === bIndex) {
-			assert(a !== undefined, "a should not be undefined if aIndex === bIndex");
-			assert(b !== undefined, "b should not be undefined if aIndex === bIndex");
-			index = a.index;
-			nodeChangeA = a.nodeChange;
-			nodeChangeB = b.nodeChange;
-			iChange += 1;
-			iOver += 1;
-		} else if (aIndex < bIndex) {
-			assert(a !== undefined, "a should not be undefined if aIndex < bIndex");
-			index = a.index;
-			nodeChangeA = a.nodeChange;
-			iChange += 1;
+		if (newIndex === baseIndex) {
+			assert(newEntry !== undefined && baseEntry !== undefined, "Entries should be defined");
+			index = newIndex;
+			newNodeChange = newEntry[1];
+			baseNodeChange = baseEntry[1];
+		} else if (newIndex < baseIndex) {
+			assert(newEntry !== undefined, "Entry should be defined");
+			index = newIndex;
+			newNodeChange = newEntry[1];
 		} else {
-			assert(b !== undefined, "b should not be undefined if aIndex > bIndex");
-			index = b.index;
-			nodeChangeB = b.nodeChange;
-			iOver += 1;
+			assert(baseEntry !== undefined, "Entry should be defined");
+			index = baseIndex;
+			baseNodeChange = baseEntry[1];
 		}
 
-		const nodeChange = rebaseChild(nodeChangeA, nodeChangeB);
+		const nodeChange = rebaseChild(newNodeChange, baseNodeChange);
 		if (nodeChange !== undefined) {
```

Comment:
Even though this call to `rebased.set(index, nodeChange);` is gated on `if (nodeChange !== undefined)`, the fact that we started off with a clone of `change` means that it will always have entries for the nested changes that it started with. Shouldn't it be possible for a nested change to dissapear due to a descendant beign moved by the based changeset?

Reply:
I see you changed the clone into a `new BTree();`. Resolving this thread.

Reply:
In practice this didn't matter because we would handle removing empty children during a separate pruning pass. But cloning the B tree was a vestige of an incremental rebasing implementation and wasn't intentional in the current implementation. I've removed it.